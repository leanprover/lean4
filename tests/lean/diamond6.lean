set_option structureDiamondWarning false

class Foo (α : Type) extends Add α where
  zero : α

class FooComm (α : Type) extends Foo α where
  comm (a b : α) : a + b = b + a

class FooAssoc (α : Type) extends Foo α, Mul α where
  add_assoc (a b c : α) : (a + b) + c = a + (b + c)
  mul_assoc (a b c : α) : (a * b) * c = a * (b * c)

class FooAC (α : Type) extends FooComm α, FooAssoc α where
  mul_comm (a b : α) : a * b = b * a

set_option pp.all true
#check FooAC.mk

#print FooAC.toFooAssoc

class FooAssoc' (α : Type) extends FooAssoc α where
  one : α

class FooAC' (α : Type) extends FooComm α, FooAssoc' α where
  mul_comm (a b : α) : a * b = b * a

#check FooAC'.mk

#print FooAC'.toFooAssoc'

def f [FooAssoc α] (a : α) : α :=
  a * a

def g [FooAC' α] (a : α) : α :=
  f (a + FooAssoc'.one)
