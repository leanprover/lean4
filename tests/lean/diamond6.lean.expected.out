FooAC.mk {α : Type} [toFooComm : FooComm α] [toMul : Mul.{0} α]
  (add_assoc :
    ∀ (a b c : α),
      @Eq.{1} α
        (@HAdd.hAdd.{0, 0, 0} α α α (@instHAdd.{0} α (@Foo.toAdd α (@FooComm.toFoo α toFooComm)))
          (@HAdd.hAdd.{0, 0, 0} α α α (@instHAdd.{0} α (@Foo.toAdd α (@FooComm.toFoo α toFooComm))) a b) c)
        (@HAdd.hAdd.{0, 0, 0} α α α (@instHAdd.{0} α (@Foo.toAdd α (@FooComm.toFoo α toFooComm))) a
          (@HAdd.hAdd.{0, 0, 0} α α α (@instHAdd.{0} α (@Foo.toAdd α (@FooComm.toFoo α toFooComm))) b c)))
  (mul_assoc :
    ∀ (a b c : α),
      @Eq.{1} α
        (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) a b) c)
        (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) a (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) b c)))
  (mul_comm :
    ∀ (a b : α),
      @Eq.{1} α (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) a b)
        (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) b a)) :
  FooAC α
def FooAC.toFooAssoc : {α : Type} → [self : FooAC α] → FooAssoc α :=
fun (α : Type) (self : FooAC α) =>
  @FooAssoc.mk α (@FooComm.toFoo α (@FooAC.toFooComm α self)) (@FooAC.toMul α self) (@FooAC.add_assoc α self)
    (@FooAC.mul_assoc α self)
FooAC'.mk {α : Type} [toFooComm : FooComm α] [toMul : Mul.{0} α]
  (add_assoc :
    ∀ (a b c : α),
      @Eq.{1} α
        (@HAdd.hAdd.{0, 0, 0} α α α (@instHAdd.{0} α (@Foo.toAdd α (@FooComm.toFoo α toFooComm)))
          (@HAdd.hAdd.{0, 0, 0} α α α (@instHAdd.{0} α (@Foo.toAdd α (@FooComm.toFoo α toFooComm))) a b) c)
        (@HAdd.hAdd.{0, 0, 0} α α α (@instHAdd.{0} α (@Foo.toAdd α (@FooComm.toFoo α toFooComm))) a
          (@HAdd.hAdd.{0, 0, 0} α α α (@instHAdd.{0} α (@Foo.toAdd α (@FooComm.toFoo α toFooComm))) b c)))
  (mul_assoc :
    ∀ (a b c : α),
      @Eq.{1} α
        (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) a b) c)
        (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) a (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) b c)))
  (one : α)
  (mul_comm :
    ∀ (a b : α),
      @Eq.{1} α (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) a b)
        (@HMul.hMul.{0, 0, 0} α α α (@instHMul.{0} α toMul) b a)) :
  FooAC' α
def FooAC'.toFooAssoc' : {α : Type} → [self : FooAC' α] → FooAssoc' α :=
fun (α : Type) (self : FooAC' α) =>
  @FooAssoc'.mk α
    (@FooAssoc.mk α (@FooComm.toFoo α (@FooAC'.toFooComm α self)) (@FooAC'.toMul α self) (@FooAC'.add_assoc α self)
      (@FooAC'.mul_assoc α self))
    (@FooAC'.one α self)
